### PR TITLE
feat: [AXM-258] add functionality to send push notifications

### PR DESCRIPTION
**Pull request by NiedielnitsevIvan**
**Description:** Add functionality to send push notifications.

**Youtrack:** https://youtrack.raccoongang.com/issue/AXM-258

**Dependencies:** will be added

**Reviewers:**
- [ ] @monteri 
- [ ] @GlugovGrGlib 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

### CODE REVIEW

File: edx_ace/__init__.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from .recipient import Recipient
 from .recipient_resolver import RecipientResolver
 
-__version__ = '1.8.0'
+__version__ = '1.8.1'
```
Comment (monteri) [inline edx_ace/__init__.py]: I suppose it is major version change or at least minor. Last number is kind of a patch like bug fix or similar

File: edx_ace/channel/__init__.py [hunk 2]
```diff
@@ -193,4 +193,15 @@ def get_channel_for_message(channel_type, message):
...
+    elif channel_type == ChannelType.PUSH:
+        try:
+            possible_channel = channels_map.get_channel_by_name(
+                channel_type,
+                settings.ACE_CHANNEL_DEFAULT_PUSH
+            )
+        except KeyError:
+            return channels_map.get_default_channel(channel_type)
+
+        return possible_channel
```
Comment (monteri) [inline edx_ace/channel/__init__.py]:
```suggestion
    if channel_type == ChannelType.EMAIL:
        if message.options.get('transactional'):
            channel_names = [settings.ACE_CHANNEL_TRANSACTIONAL_EMAIL, settings.ACE_CHANNEL_DEFAULT_EMAIL]
        else:
            channel_names = [settings.ACE_CHANNEL_DEFAULT_EMAIL]
    elif channel_type == ChannelType.PUSH:
        channel_names = [settings.ACE_CHANNEL_DEFAULT_PUSH]
    possible_channels = []
    try:
        possible_channels = [
            channels_map.get_channel_by_name(channel_type, channel_name)
            for channel_name in channel_names
            ]
    except KeyError:
        return channels_map.get_default_channel(channel_type)
    # First see if any channel specifically demands to deliver this message
    for channel in possible_channels:
         if channel.overrides_delivery_for_message(message):
             return channel
        # Else the normal path: use the preferred channel for this message type
        return possible_channels[0] if possible_channels else channels_map.get_default_channel(channel_type)
```

Comment (monteri) [reply edx_ace/channel/__init__.py]: Can we consider this refactor? I assume that
```
if channel_type == ChannelType.EMAIL:
        if message.options.get('transactional'):
            channel_names = [settings.ACE_CHANNEL_TRANSACTIONAL_EMAIL, settings.ACE_CHANNEL_DEFAULT_EMAIL]
        else:
            channel_names = [settings.ACE_CHANNEL_DEFAULT_EMAIL]
    elif channel_type == ChannelType.PUSH:
        channel_names = [settings.ACE_CHANNEL_DEFAULT_PUSH]
```
can be separate from other logic where the `possible_channel` is defined

File: edx_ace/channel/push_notification.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            return
+
+        for token in device_tokens:
+            self.send_message(token, rendered_message)
+
+    def send_message(self, token: str, rendered_message: RenderedPushNotification) -> None:
+        """
+        Send a push notification to a device by token.
+        """
+        notif_data = {
```
Comment (monteri) [inline edx_ace/channel/push_notification.py]:
```suggestion
        notification_data = {
```

File: edx_ace/renderers.py [hunk 4]
```diff
@@ -81,3 +82,21 @@ class EmailRenderer(AbstractRenderer):
...
+
+@attr.s
+class RenderedPushNotification:
+    """
+    Encapsulates all values needed to send a :class:`.Message`
+    over an :attr:`.ChannelType.PUSH`.
+    """
+
+    subject = attr.ib()
+    body = attr.ib()
```
Comment (monteri) [inline edx_ace/renderers.py:94]: Should there be `data` attribute that represents additional dictionary data?

**Review by monteri — COMMENTED** [review]
LGTM, left several nit comments

File: edx_ace/renderers.py (hunk 4 above)
Comment (NiedielnitsevIvan) [reply edx_ace/renderers.py:94]: No, the renderer has a different purpose: it responds to templates for different parts of notifications/emails, and the context for them is in the message object and is inserted into the templates from there.

File: edx_ace/__init__.py (hunk 1 above)
Comment (NiedielnitsevIvan) [reply edx_ace/__init__.py]: Done

File: edx_ace/channel/__init__.py (hunk 2 above)
Comment (NiedielnitsevIvan) [reply edx_ace/channel/__init__.py]: Refactored

**Review by monteri — APPROVED** [review]

**Pull request merged**